### ISSUE TITLE
Add monitors to auto-download tracked searches

### ISSUE THREAD

**Issue by Bionus** [Bionus/imgbrd-grabber#1620]
## Summary
Add a new "special" tab to the application (along with "Favorites", "Downloads", and "Log"): "Monitors".

## Rationale
Currently, favorites can be monitored to trigger a notification through the application's system tray icon.
This proposal is to expand on this, and extract the monitoring logic from the favorites.

## Description
A monitor would be composed of:
* A search: a tag or a gallery
* Sources: one ore more
* An interval: how often the monitor will check for new images
* Actions: what to do when new images arrive (zero or more)

I see a two possible actions:
* Notify: trigger a notification via the tray icon, same as what you can do with favorites currently
* Download: downlad the images, with or without an override of the path/filename

The UI would be a new tab with a table listing the monitors, with the "search" column being a clickable link/button that would open the "search" in a new tab.

-----

Any comments? Additional suggestion? Arguments for/against?
I'd love some feedback on this :smile:

**Comment by Bionus** [comment]
The main issue I have with this is how the monitoring logic in the favorites should be impacted.

Should the "last read" / red circle (favorites monitoring) feature be removed in favor of actual monitors?
⚠️ This would require users to duplicate their queries in both favorites and monitors, which can be painful.

Should the monitor logic merge with the current "favorites" tab, with actual "favorites" being simplified as the "read for later" tabs?
⚠️ This would greatly reduce the visibility of favorites.

**Comment by MrAndre96** [comment]
For the automatic download how do you plan to introduce it? Will you download it on the background or will it open after a notification a batch download prompt and how will it handle if multipele tags have an update? Will it directly continue to download the second after the first tag is completed or will you get a notification between the first tag and the second tag etc

For the favorites and monitoring duplicate you could maybe check if an tag is already in one or another so you would only be notified once for that tag and not twice.

**Comment by Bionus** [comment]
> For the automatic download how do you plan to introduce it? Will you download it on the background or will it open after a notification a batch download prompt and how will it handle if multipele tags have an update? Will it directly continue to download the second after the first tag is completed or will you get a notification between the first tag and the second tag etc

I was planning of simply downloading them in the background, one after the other. The notifications would all pop as soon as the update is detected however, **not** when the download starts.

But now that you mention it it also means that the user cannot really see if any image was "ignored" or this kind of thing. 🤔 

> For the favorites and monitoring duplicate you could maybe check if an tag is already in one or another so you would only be notified once for that tag and not twice.

My main issue is more to have different equivalent logic ("monitor a search") in two different places of the UI, which is confusing for the user and annoying to maintain for me. For example, what if someone is tracking a favorite (for the red circle) and now wants to add auto-download? That would be in another tab, which isn't very user-friendly.

But on the other hand, having monitoring integrated in the favorites tab allows for some "favorites-only" features such as the red circle with the number of images unseen since last time.

The only "correct" option I can think of is adding a simple checkbox in the favorite's setting, which when clicked would generate a monitor in the "Monitors" tab with notifications enabled. Then, on the favorites tab, if any monitor exist with the same tags as the favorite, show the relevant red circle.

**Comment by MrAndre96** [comment]
> I was planning of simply downloading them in the background, one after the other. The notifications would all pop as soon as the update is detected however, **not** when the download starts.
> 
> But now that you mention it it also means that the user cannot really see if any image was "ignored" or this kind of thing. 🤔

How would you let the user know that something is downloading in the background? (For eample when you want to close grabber). 

Also how would it work when you then can have a download in the background and in the front with for example a batch download?

**Comment by Bionus** [comment]
> How would you let the user know that something is downloading in the background? (For example when you want to close grabber).

Indeed, while it would obviously be visible directly from the "Monitors" tab, it should also be visible from other places. I guess either in the tab name (some icon to show something's happening), in the window's title, or by adding a status bar (🤢).

> Also how would it work when you then can have a download in the background and in the front with for example a batch download?

Very good point, especially if the "simultaneous downloads" is set to 1, or for annoying sources that limit downloads to one per X seconds. I was personally thinking of simply having no simultaneous downloads for the monitors (i.e. one image at a time) to avoid taking too much bandwidth when the user is doing other things, but haven't thought about batch downloads.

I guess the easiest solution would be to pause all background downloads when a batch download starts, and resume them when it's done.

Unless a big revamp of the batch downloads is done, where you could have multiple batch downloads in parallel (for different sources).

**Comment by MelanieTanaka** [comment]
Honestly the only reason I don't use this app after having tried it is that it doesn't auto-download. When I found this I was expecting behavior similar to hdoujindownloader, but for images instead of doujin/manga. I have a separate web app for viewing files, so I want to just set this app on a virtual machine and have it feed the folders my other apps use to display the content. i.e: other than adding new tags I don't want to ever touch it.

**Comment by Bionus** [comment]
Missing features:
* [ ] Human-friendly interval display (currently raw seconds)
* [ ] Multiple sources
* [ ] Post-filter
* [ ] Enable/disable notifications per monitor

Getting rid of favorites' monitors might require a migration so I'll leave it as out-of-scope for this story. Both might coexist in the next release in case there are issues with the new monitors.

**Comment by Triramama** [comment]
so how do i use "monitor" function exactly? don't think it's described anywhere on the faq site
